Make utility for generating git url string

omus opened the pull request:
Part of #20725.

Various parts of the callback code create URLs for displaying information to the user. The `git_url` function allows us to consistently create these URL strings and allows us to test behaviour.

omus committed:
Make utility for generating git url string

omus commented on base/libgit2/callbacks.jl:
```diff
@@ -224,7 +229,7 @@ function credentials_callback(libgit2credptr::Ptr{Ptr{Void}}, url_ptr::Cstring,
     if isset(allowed_types, Cuint(Consts.CREDTYPE_USERPASS_PLAINTEXT))
         defaultcreds = reset!(UserPasswordCredentials(true), -1)
-        credid = "$schema$host"
+        credid = "$(isempty(schema) ? "ssh" : schema)://$host"
```
Slight change to behaviour here. It used to be that when the `schema` was not defined in the URL the credential ID would be `"$host"` now it will be `"ssh://$host"`.

tkelman replied:
I'm not sure defaulting to ssh is any better than requiring setting a schema

omus replied:
Note that using the scp-like syntax `git@github.com:JuliaLang/julia.git` or `ssh://git@github.com/JuliaLang/julia.git` both refer to the same git repo using the SSH protocol. 

The advantage of defaulting to using SSH here is that both of these syntaxes will refer to the same `credid`.

See https://git-scm.com/book/id/v2/Git-on-the-Server-The-Protocols (under "The SSH Protocol" section)

omus replied:
We could just assign "ssh" to be the schema when one is missing but this would then modify the `git_url` output from `git@github.com` to `ssh://git@github.com` which would no longer resemble the URL passed in by the user.

tkelman replied:
what would happen on `git@github.com/JuliaLang/julia.git` ?

omus replied:
Based upon the behavour of `git clone` that URL should be invalid.

tkelman replied:
right but if it gets passed here will you be prepending `ssh` to it?

omus replied:
Currently yes. I'll update the URL_REGEX to fix this.

omus replied:
URL_REGEX no longer allows `git@github.com/JuliaLang/julia.git`

yuyichao commented:
Just want to make sure this happens **after** alias resolution in `.gitconfig` right?

omus commented:
> Just want to make sure this happens **after** alias resolution in `.gitconfig` right?

I believe so. What is your concern?

yuyichao commented:
Just to make sure it's not broken.

omus committed:
Disallow scp-like URLs not using : as separator

Git URLs like `git@github.com:repo` are valid while URLs like
`git@github.com/repo` are now treated as wrong.

omus committed:
Disallow HTTP URLs using : as path separator

Previously URLs like `http://git@github.com:repo` were being parsed
incorrectly.

tkelman commented on test/libgit2.jl:
```diff
@@ -183,15 +183,21 @@ end
     @testset "HTTPS URL, no path" begin
         m = match(LibGit2.URL_REGEX, "https://user:pass@server.com:80")
-        @test m[:path] == ""
+        @test m[:path] == nothing
```
`===`

tkelman commented on base/libgit2/utils.jl:
```diff
@@ -3,7 +3,7 @@
 # https://git-scm.com/docs/git-clone#_git_urls_a_id_urls_a
 const URL_REGEX = r"""
-^(?:(?<scheme>ssh|git|https?)://)?
+^(?:(?<scheme>ssh|git|https?)://)?+
```
what's the difference with this? regex101.com tells me it's greedy vs possessive, whether or not it "gives back" but I'm not familiar with the consequences of that

omus replied:
Basically when `https://` is matched the regular expression will stop any backtracking where this string could be repurposed elsewhere in the regular expression:

```julia
julia> match(possessive_regex, "https://git@server:repo")

julia> match(original_regex, "https://git@server:repo")
RegexMatch("https://git@server:repo", scheme=nothing, user="https", password="//git", host="server", port=nothing, path="repo")
```

tkelman replied:
oh ick. that's subtle, good catch and thanks for the explanation

omus committed:
[ci skip] Add docstring for git_url

omus commented:
Added some documentation. Ready to merge.

tkelman commented on base/libgit2/utils.jl:
```diff
@@ -70,6 +70,38 @@ else Sys.isunix()
+  * `path::AbstractString=""`: the path to specify to use in the output if provided.
+
+# Examples
+
```
no blank line here

omus replied:
Should I remove the blank line after `# Keywords` as well?

tkelman commented on base/libgit2/utils.jl:
```diff
@@ -70,6 +70,38 @@ else Sys.isunix()
+    be specified.
+  * `port::Union{AbstractString,Integer}=""`: the port number to use in the output if
+    provided. Cannot be specified when using the scp-like syntax.
+  * `path::AbstractString=""`: the path to specify to use in the output if provided.
```
"to specify to use" sounds a bit redundant

omus replied:
Weird. I don't remember writing it like that. Will fix.

omus commented:
I think this is ready to go. Will merge on Tuesday.
